bpf/ingress_node_firewall_kernel.c

```diff
@@ -444,21 +444,83 @@ ingress_node_firewall_main(struct xdp_md *ctx) {
     switch (action) {
     case DENY:
         generate_event_and_update_statistics(ctx, bpf_xdp_get_buff_len(ctx), DENY, ruleId, 1, ifId);
-        ingress_node_firewall_printk("Ingress node firewall action DENY -> XDP_DROP");
+        ingress_node_firewall_printk("XDP: Ingress node firewall action DENY -> XDP_DROP");
         return XDP_DROP;
     case ALLOW:
         generate_event_and_update_statistics(ctx, bpf_xdp_get_buff_len(ctx), ALLOW, ruleId, 0, ifId);
-        ingress_node_firewall_printk("Ingress node firewall action ALLOW -> XDP_PASS");
+        ingress_node_firewall_printk("XDP: Ingress node firewall action ALLOW -> XDP_PASS");
         return XDP_PASS;
     default:
-        ingress_node_firewall_printk("Ingress node firewall action UNDEF");
+        ingress_node_firewall_printk("XDP: Ingress node firewall action UNDEF");
         return XDP_PASS;
     }
 }
 
+/*
+ * tcx_ingress_node_firewall_main(): is the entry point for the TCX program to do
+ * ingress node firewall.
+ * Input:
+ * struct __sk_buff *ctx: pointer to sk_buff which contains packet pointer and input interface index.
+ * Output:
+ * none.
+ * Return:
+ * int TCX action: valid values TCX_DROP and TCX_NEXT.
+ */
+__attribute__((__always_inline__)) static inline int
+tcx_ingress_node_firewall_main(struct __sk_buff *ctx) {
+    void *data = (void *)(long)ctx->data;
+    void *dataEnd = (void *)(long)ctx->data_end;
+    struct ethhdr *eth = data;
+    void *dataStart = data + sizeof(struct ethhdr);
+    __u32 result = UNDEF;
+    __u32 ifId = ctx->ifindex;
+
+    ingress_node_firewall_printk("TCX: Ingress node firewall start processing a packet on %d", ifId);
+
+    if (unlikely(dataStart > dataEnd)) {
+        ingress_node_firewall_printk("TCX: Ingress node firewall bad packet TCX_DROP");
+        return TCX_DROP;
+    }
+    switch (eth->h_proto) {
+    case bpf_htons(ETH_P_IP):
+        ingress_node_firewall_printk("TCX: Ingress node firewall process IPv4 packet");
+        result = ipv4_firewall_lookup(data, dataEnd, ifId);
+        break;
+    case bpf_htons(ETH_P_IPV6):
+        ingress_node_firewall_printk("TCX: Ingress node firewall process IPv6 packet");
+        result = ipv6_firewall_lookup(data, dataEnd, ifId);
+        break;
+    default:
+        ingress_node_firewall_printk("TCX: Ingress node firewall unknown L3 protocol TCX_NEXT");
+        return TCX_NEXT;
```

Comment:
If we have an unknown proto, what does `TCX_NEXT` mean? 
I see:
```
+/* (Simplified) user return codes for tcx prog type.
+ * A valid tcx program must return one of these defined values. All other
+ * return codes are reserved for future use. Must remain compatible with
+ * their TC_ACT_* counter-parts. For compatibility in behavior, unknown
+ * return codes are mapped to TCX_NEXT.
+ */
+enum tcx_action_base {
+	TCX_NEXT	= -1,
+	TCX_PASS	= 0,
+	TCX_DROP	= 2,
+	TCX_REDIRECT	= 7,
+};
```

So it returns unknown? And what happens then?

Reply:
in case of chained TCX hooks it will go and run the next one if none it will be the same effect as `TCX_PASS`